libzmix/src/utils/bls12_381/mod.rs

```diff
@@ -0,0 +1,237 @@
+                    res.append(&mut vec);
+                    vec = res;
+                }
+                $name($internal::frombytes(vec.as_slice()))
+            }
+        }
+
+        impl serde::ser::Serialize for $name {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                serializer.serialize_newtype_struct(stringify!($name), &self.to_string())
+            }
+        }
+
+        impl<'a> serde::de::Deserialize<'a> for $name {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: serde::de::Deserializer<'a>,
+            {
+                struct $visitor;
+
+                impl<'a> serde::de::Visitor<'a> for $visitor {
+                    type Value = $name;
+
+                    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                        formatter.write_str(stringify!($name))
+                    }
+
+                    fn visit_str<E>(self, value: &str) -> Result<$name, E>
+                    where
+                        E: serde::de::Error,
+                    {
+                        let name_str = stringify!($name);
+                        if value.starts_with(name_str) {
+                            Ok(
+                                $name::from_str(&value[(name_str.len() + 2)..(value.len() - 2)])
+                                    .map_err(serde::de::Error::custom)?,
+                            )
+                        } else {
+                            Err(E::custom(format!("Invalid string: {}", value)))
+                        }
+                    }
+                }
+
+                deserializer.deserialize_str($visitor)
+            }
+        }
+    };
+}
+
+macro_rules! format_impl {
+    ($name:ident) => {
+        impl std::fmt::Display for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "{}( {} )", stringify!($name), self.to_hex())
+            }
+        }
+
+        impl std::fmt::Debug for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "{}( {} )", stringify!($name), self.to_hex())
+            }
+        }
+    };
+}
+
+macro_rules! to_bytes {
+    () => {
+        pub fn to_bytes(&self) -> Vec<u8> {
+            let mut res = vec![0u8; Self::BYTES_REPR_SIZE];
+            self.repr_bytes(&mut res);
+            res
+        }
+    };
+}
+
+macro_rules! constructor {
+    ($name:ident) => {
+        pub fn new() -> Self {
+            $name(random_mod_order::<ThreadRng>(None))
+        }
+    };
+}
+
+/// Generate a random number twice the number of bytes as the curve reduced by mod CURVE_ORDER
+/// Here 96 bytes are randomly generated mod CURVE_ORDER. The result is 48 bytes.
+fn random_mod_order<R: RngCore>(r: Option<&mut R>) -> BIG {
```

Comment:
Why is it generating twice the amount of entropy needed (`compute_big`)?

Reply:
To make sure there is no bias is the randomization. Recommendation from @hartm

Reply:
Can you or @hartm provide a reference for this kind of extractor? And do we need a randomness extractor. I thought the entropy source on the machine (assuming you are using a standard machine) should remove any statistical bias.

Reply:
The Curve order is not exactly the maximum value for 2^381 right. So if you extract 384 bytes you don't get exactly an even distribution and hence the bias towards the lower numbers.